prusti-viper/src/encoder/procedure_encoder.rs

```diff
@@ -4985,32 +5054,78 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
         self.encode_copy_value_assign(encoded_lhs, encoded_val, ty, location)
     }
 
-    fn encode_assign_array_len(
+    fn encode_assign_sequence_len(
         &mut self,
         encoded_lhs: vir::Expr,
         place: &mir::Place<'tcx>,
         dst_ty: ty::Ty<'tcx>,
         location: mir::Location,
     ) -> SpannedEncodingResult<Vec<vir::Stmt>> {
         trace!(
-            "[enter] encode_assign_array_len(place={:?}, ty={:?})",
+            "[enter] encode_assign_sequence_len(place={:?}, ty={:?})",
             place,
             dst_ty,
         );
         let span = self.mir_encoder.get_span_of_location(location);
+        let (encoded_place, mut stmts, place_ty, ..) = self.encode_place(place).with_span(span)?;
+        let ty_name = self.encoder.encode_type_predicate_use(place_ty)
+            .with_span(span)?;
+        match place_ty.kind() {
+            ty::TyKind::Array(_, ref ty_len) => {
+                // extract the length from the array type
+                let len = self.encoder.const_eval_intlike(&ty_len.val).unwrap()
+                    .to_u64().unwrap();
+
+                stmts.extend(
+                    self.encode_copy_value_assign(
+                        encoded_lhs,
+                        len.into(),
+                        dst_ty,
+                        location,
+                    )?
+                );
+            },
+            ty::TyKind::Slice(..) => {
+                let (slice_ty_pred, _slice_ty, elem_ty_pred, _elem_ty, _elem_value_ty, _elem_ty_rs) =
+                    self.encode_slice_types(place_ty)
+                        .with_span(span)?;
 
-        let place_ty = self.encode_place(place).with_span(span)?.2;
-        let ty_len = if let ty::TyKind::Array(_, ref ty_len) = place_ty.kind() { ty_len } else { unreachable!() };
+                // encode Slice$<elem_ty>$len
+                let slice_len = self.encoder.encode_builtin_function_use(
+                    BuiltinFunctionKind::SliceLen {
+                        slice_ty_pred,
+                        elem_ty_pred,
+                    }
+                );
 
-        let len = self.encoder.const_eval_intlike(&ty_len.val).unwrap()
-            .to_u64().unwrap();
+                stmts.push(
+                    vir::Stmt::Assign(
+                        self.encoder.encode_value_expr(encoded_lhs, dst_ty),
+                        vir::Expr::func_app(
+                            slice_len,
+                            vec![
+                                encoded_place,
+                            ],
+                            vec![
+                                vir::LocalVar::new(
+                                    String::from("self"),
+                                    vir::Type::TypedRef(ty_name),
+                                ),
```

Comment:
You use this a lot, so it may be better to define a method such that you can just call `vir::LocalVar::new_typed_ref("self", ty_name)`.